### PR TITLE
semantic_text field mapping

### PR DESCRIPTION
**Pull request by carlosdelest**
This PR contains changes related to just the `semantic_text` field mapping, including some specific YAML mapping tests.

Automatic Inference on ingestion will be included as a separate PR. This change deals with just the mapping side:
  - Provides the mapping definition for `semantic_text`
  - On document ingestion, checks for the model settings in its mapping. If it's not present, updates it to ensure documents adhere to the mapping in the future.
  - Creates the appropriate mappings for `dense_vector` or `sparse_vector` fields internally
  - Validates and indexes the inference results as part of the document ingestion using these internal mappings

### CODE REVIEW

**Review by benwtrent — COMMENTED** [review]
First pass complete. Need to go through the tests now :)

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+        });
+
+        private final Parameter<SemanticTextField.ModelSettings> modelSettings = new Parameter<>(
+            "model_settings",
+            true,
+            () -> null,
+            (n, c, o) -> SemanticTextField.parseModelSettingsFromMap(o),
+            mapper -> ((SemanticTextFieldType) mapper.fieldType()).modelSettings,
+            XContentBuilder::field,
+            (m) -> m == null ? "null" : Strings.toString(m)
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java]: I think we were trying to avoid defining `toString()` here but reusing the `Strings.toString()` and using `toXContent()` to render the conflicts message. 

I don't see that we're doing that anywhere else, so I'm implementing `toString()` on the `ModelSettings` class and updating the tests accordingly.

@jimczi , LMK if I'm missing anything here!

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+            case SPARSE_EMBEDDING -> new SparseVectorFieldMapper.Builder(CHUNKED_EMBEDDINGS_FIELD.getPreferredName());
+            case TEXT_EMBEDDING -> {
+                DenseVectorFieldMapper.Builder denseVectorMapperBuilder = new DenseVectorFieldMapper.Builder(
+                    CHUNKED_EMBEDDINGS_FIELD.getPreferredName(),
+                    indexVersionCreated
+                );
+                SimilarityMeasure similarity = modelSettings.similarity();
+                if (similarity != null) {
+                    switch (similarity) {
+                        case COSINE -> denseVectorMapperBuilder.similarity(DenseVectorFieldMapper.VectorSimilarity.COSINE);
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:360]: It seems this got added after we were working on this part of the code. I'm adding that, thanks for the catch! 🙇

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        }
+        if (previous == null) {
+            return true;
+        }
+        if (current == null) {
+            conflicts.addConflict("model_settings", "");
+            return false;
+        }
+        conflicts.addConflict("model_settings", "");
+        return false;
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java]: Makes no sense, I'm simplifying it

File: x-pack/plugin/inference/src/yamlRestTest/java/org/elasticsearch/xpack/inference/InferenceRestIT.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import org.junit.ClassRule;
+
+public class InferenceRestIT extends ESClientYamlSuiteTestCase {
+
+    @ClassRule
+    public static ElasticsearchCluster cluster = ElasticsearchCluster.local()
+        .setting("xpack.security.enabled", "false")
+        .setting("xpack.security.http.ssl.enabled", "false")
+        .plugin("inference-service-test")
+        .distribution(DistributionType.DEFAULT)
```
Comment (mark-vieira) [inline x-pack/plugin/inference/src/yamlRestTest/java/org/elasticsearch/xpack/inference/InferenceRestIT.java:25]: Is it strictly necessary we use the default distribution here?

Comment (carlosdelest) [reply x-pack/plugin/inference/src/yamlRestTest/java/org/elasticsearch/xpack/inference/InferenceRestIT.java]: No - we can use the integration test distribution adding another plugin. Thanks for the catch!

**Comment by carlosdelest** [comment]
@elasticsearchmachine update branch

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java [hunk 5]
```diff
@@ -696,6 +696,10 @@ private static void failIfMatchesRoutingPath(DocumentParserContext context, Stri
      */
     private static void parseCopyFields(DocumentParserContext context, List<String> copyToFields) throws IOException {
         for (String field : copyToFields) {
+            if (context.mappingLookup().getMapper(field) instanceof InferenceFieldMapper) {
+                // ignore copy_to that targets inference fields, values are already extracted in the coordinating node to perform inference.
+                continue;
+            }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:702]: drive-by comment: could we try and avoid this type of instanceof checks / special cases? What would it take to remove it? Would it be worthwhile to generalize the special needs of inference field mapper into Mapper somehow? Other ideas? I think it's important that we at least consider it.

File: server/src/main/java/org/elasticsearch/index/mapper/Mapping.java [hunk 6]
```diff
@@ -76,7 +76,7 @@ public CompressedXContent toCompressedXContent() {
     /**
      * Returns the root object for the current mapping
      */
-    RootObjectMapper getRoot() {
+    public RootObjectMapper getRoot() {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/Mapping.java:79]: drive-by comment: where is this needed to become public?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java]: We can define `toString` but this part is about serialising the object in case of conflict so we need to handle `null` values. To simplify things we can define toString on the `ModelSettings` class (using Strings.toString(this) should be fine) but we have to use `Objects.toString(m)` to handle null values. 
Not that `Objects#toString` handles `null` values by returning "null" so it would be equivalent to what we're doing here.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+        MODEL_SETTINGS_PARSER.declareString(ConstructingObjectParser.constructorArg(), TASK_TYPE_FIELD);
+        MODEL_SETTINGS_PARSER.declareInt(ConstructingObjectParser.optionalConstructorArg(), DIMENSIONS_FIELD);
+        MODEL_SETTINGS_PARSER.declareString(ConstructingObjectParser.optionalConstructorArg(), SIMILARITY_FIELD);
+    }
+
+    /**
+     * Converts the provided {@link ChunkedInferenceServiceResults} into a list of {@link Chunk}.
+     */
+    public static List<Chunk> toSemanticTextFieldChunks(
```
[the remark this answers is not in the document]
Comment (jimczi) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java]: This is also used by the `ShardBulkInferenceActionFilter` to translate the response from the inference call.
This class is not in this PR so I am fine if we move it to tests for now but it will need to move back when we introduce the action filter.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+                        );
+                    }
+                    if (similarity == null) {
+                        throw new IllegalArgumentException(
+                            "required [" + SIMILARITY_FIELD + "] field is missing for task_type [" + taskType.name() + "]"
+                        );
+                    }
+                    break;
+                case SPARSE_EMBEDDING:
+                    break;
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:122]: > I guess that would make us more forward compatible in case this is used in a mixed cluster env, but I could be wrong. Should we be more strict for non supported params?

But then the old mappings wouldn't even be using those values and consequently, we end up with things indexed very differently in a mixed cluster environment? 

I think we should be strict here. We shouldn't allow parameters that are invalid.

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java [hunk 9]
```diff
@@ -696,6 +696,10 @@ private static void failIfMatchesRoutingPath(DocumentParserContext context, Stri
      */
     private static void parseCopyFields(DocumentParserContext context, List<String> copyToFields) throws IOException {
         for (String field : copyToFields) {
+            if (context.mappingLookup().inferenceFields().get(field) != null) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:699]: Is this even something we should be checking at runtime? Shouldn't we just remove copy_to that targets inference fields from the mapping that the parser works with up-front so we save this cost?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+ * @param fieldName The original field name.
+ * @param originalValues The original values associated with the field name.
+ * @param inference The inference result.
+ * @param contentType The {@link XContentType} used to store the embeddings chunks.
+ */
+public record SemanticTextField(String fieldName, List<String> originalValues, InferenceResult inference, XContentType contentType)
+    implements
+        ToXContentObject {
+
+    static final ParseField TEXT_FIELD = new ParseField("text");
```
Comment (original-brownbear) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java:54]: Doesn't it make more sense for all of these to be `String` and then just build the one-time use `ParseField` instances when you set up the parser? It would certainly make the code easier to read by not having all the `getPreferredName()` calls there and for the string concatenations would even compile to better code.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java [hunk 11]
```diff
@@ -0,0 +1,470 @@
...
+                        + " is not using different model settings",
+                    exc
+                );
+            }
+            mapper = this;
+        }
+        var chunksField = mapper.fieldType().getChunksField();
+        var embeddingsField = mapper.fieldType().getEmbeddingsField();
+        for (var chunk : field.inference().chunks()) {
+            XContentParser subParser = XContentHelper.createParserNotCompressed(
```
Comment (original-brownbear) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:240]: It's good to close parsers after you're done with them, saves some memory from pooled resources managed by Jackson :)

**Review by original-brownbear — COMMENTED** [review]
Didn't get through all of this yet but left a couple of things that should get fixed IMO. I'll continue later

File: server/src/test/java/org/elasticsearch/index/mapper/CopyToMapperTests.java [hunk 12]
```diff
@@ -22,6 +22,7 @@
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
```
[the remark this answers is not in the document]
Comment (carlosdelest) [reply server/src/test/java/org/elasticsearch/index/mapper/CopyToMapperTests.java:25]: Makes sense. I've added them in a new test class - `MappingLookupInferenceFieldMapperTests`. LMKWYT!

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java (hunk 9 above)
Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:699]: I see your point. Maybe changing the actual field mapping would be too much - how about adding a new structure on the field type lookup for dealing with the copy_to list of fields to parse?

I've done a draft on [this PR](https://github.com/carlosdelest/elasticsearch/pull/14), WDYT?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java (hunk 10 above)
Comment (carlosdelest) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextField.java]: It does - thanks for the proposal (11d6b3caa8243ee10fe2b3f000b33a41dabefbbd).

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java (hunk 11 above)
Comment (carlosdelest) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java]: Sure - let me add that, thanks! (f8dfd9c24ea789c52bc2c6110e04e40f51cae928)

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java (hunk 9 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:699]: Is it really about cost? I mean it should be transparent performance wise. I don't think we should add a new structure in the lookup just for this.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:699]: Yea lets not add a new structure, anything that increases the size of these things makes me uneasy after many-shards :) I was just thinking, if we know all the fields we don't want to copy_to upfront, why not just filter the `copyToFieldsList` at parser build time? Is that really something that is hard to do?

But on second thought! IS this even ok? Do we even want to allow this kind of mapping conflict? Shouldn't we just not have a copy_to definition that targets a field that gets populated by this mapper to begin with?

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:699]: We want to allow copy_to to target semantic text fields and we need to access this information somewhere so that we can do the actual copy in the coordinated node (when the inference is applied). The only difference here is on where we apply the copy_to. For semantic text fields it happens on the coordinating node, for other fields it happens during the parsing on the shard.

Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:699]: > why not just filter the copyToFieldsList at parser build time? Is that really something that is hard to do?

Do you mean to use the DocumentParserContext.copyToFields() instead of the copyTo() information in the mapping? AFAIU this list is used for a different purpose.

If we're not using a new structure in field type lookup, I guess the two available options would be:
-Keep track of the effective copy_to list in the parser as an additional param to its methods, which will be cumbersome but doable
- Create a copy and change the underlying mappings copy_to, which is probably overkill

Is there another option that I'm not seeing here?

> Shouldn't we just not have a copy_to definition that targets a field that gets populated by this mapper to begin with?

Having copy_to targeting a semantic_text field is a valid use case - it's just that we don't want the parser to do the parsing, as the inference process will have already populated the inference results using all sources, so  the semantic text field mapper must not be invoked separately from the source fields.

Comment (carlosdelest) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:699]: After talking offline with @original-brownbear , we're parking this for now and will optimize later if needed 👍

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/10_semantic_text_field_mapping.yml [hunk 13]
```diff
@@ -0,0 +1,105 @@
+setup:
+  - skip:
+      version: " - 8.12.99"
+      reason: semantic_text introduced in 8.13.0 # TODO change when 8.13.0 is released
```
Comment (benwtrent) [inline x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/10_semantic_text_field_mapping.yml:4]: This should be bumped.

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/20_semantic_text_field_mapping_incompatible_field_mapping.yml [hunk 14]
```diff
@@ -0,0 +1,165 @@
+setup:
+  - skip:
+      version: " - 8.12.99"
+      reason: semantic_text introduced in 8.13.0 # TODO change when 8.13.0 is released
```
Comment (benwtrent) [inline x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/20_semantic_text_field_mapping_incompatible_field_mapping.yml:4]: This should be bumped

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/10_semantic_text_field_mapping.yml [hunk 15]
```diff
@@ -0,0 +1,105 @@
+setup:
```
Comment (benwtrent) [inline x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/10_semantic_text_field_mapping.yml:1]: Could you add some tests for the following cases? Yaml tests give us a ton more coverage.

 - Multi-field (itself as a multi-field and multi-fields UNDER it)
 - nested fields
 - copy to
 - Missing inference chunk text, missing inference results (you already have good tests around bad params or mixed types, huzzah!)

**Review by benwtrent — COMMENTED** [review]
Some comments on the tests. But I think this is almost ready to go :)

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/20_semantic_text_field_mapping_incompatible_field_mapping.yml (hunk 14 above)
Comment (carlosdelest) [reply x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/20_semantic_text_field_mapping_incompatible_field_mapping.yml:4]: 👍

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/10_semantic_text_field_mapping.yml (hunk 13 above)
Comment (carlosdelest) [reply x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/10_semantic_text_field_mapping.yml:4]: Done, thanks!

**Review by benwtrent — APPROVED** [review]

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/10_semantic_text_field_mapping.yml (hunk 15 above)
Comment (benwtrent) [reply x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/10_semantic_text_field_mapping.yml:1]: the copy-to check will be interesting, we need to be sure to handle it correctly :)

**Comment by carlosdelest** [comment]
@elasticsearchmachine update branch

**Comment by carlosdelest** [comment]
@elasticmachine update branch

**Comment by carlosdelest** [comment]
@elasticmachine update branch
